diff --git a/CONSTANTS.jl b/CONSTANTS.jl
index 8714646..992fc09 100644
--- a/CONSTANTS.jl
+++ b/CONSTANTS.jl
@@ -5,7 +5,7 @@
 # 
 # Eryn Cangi
 # Created November 2021
-# Currently tested for Julia: 1.8.5
+# Currently tested for Julia: 1.11.2
 ################################################################################
 
 #                              Universal constants
@@ -14,7 +14,7 @@ const kB_MKS = 1.38e-23;        # J/K - needed for saturation vapor pressure emp
 const kB = 1.38e-16;            # erg/K
 const bigG = 6.67e-8;           # dyne-cm^2/g^2
 const mH = 1.67e-24;            # g 
-const q = -4.8032e-10            # statcoulomb (cm^1.5 g^0.5 s^-1)
+const q = -4.8032e-10;            # statcoulomb (cm^1.5 g^0.5 s^-1)
 const density_water_cgs = 0.997408  # used mostly for conversions
 const s_per_yr = 3.154e7
 const SMOW = 1.6e-4             # Standard mean ocean water
@@ -22,7 +22,7 @@ const SMOW = 1.6e-4             # Standard mean ocean water
 # Polarizability from NIST. Experimental values from: https://cccbdb.nist.gov/pollistx.asp
 # Calculations for species not available in experiment from: https://cccbdb.nist.gov/polcalc2x.asp
 # Deuterated species not listed in either are estimated by me to be the same as their H-bearing analogue.
-# I used the calcualtions that use "Density functional", "aug-cc-PVDZ", and "mPW1PW91" 
+# I used the calculations that use "Density functional", "aug-cc-PVDZ", and "mPW1PW91" 
 # because that was the method that gave the closest answer for HD to the experimental value. 
 # I have no idea what any of it means or whether it's reasonable. I'm not a quantum chemist.
 # Values are given in cm^3
