diff --git a/Photochemistry/src/Photochemistry.jl b/Photochemistry/src/Photochemistry.jl
index 1b03b28..55d47fe 100644
--- a/Photochemistry/src/Photochemistry.jl
+++ b/Photochemistry/src/Photochemistry.jl
@@ -82,7 +82,8 @@ export
     fluxcoefs, 
     Keddy, 
     update_diffusion_and_scaleH, 
-    update_transport_coefficients,  
+    update_transport_coefficients,
+    update_horiz_transport_coefficients,
 
     ## Water
     precip_microns, 
@@ -117,10 +118,11 @@ export
     get_counts,
     getpos, 
     input, 
-    logrange, 
-    nans_present, 
-    next_in_loop, 
-    searchsortednearest, 
+    logrange,
+    nans_present,
+    arrays_equal_with_nan,
+    next_in_loop,
+    searchsortednearest,
 
     ## String Manipulation
     decompose_chemistry_string, 
